### PR TITLE
Kv secrets

### PR DESCRIPTION
**Pull request by ianhelle**
Keyvault/keyring protected secrets for msticpyconfig.yaml

### CODE REVIEW

**Comment by ianhelle** [comment]
## Summary of changes:
- keyvault_secrets.py and secret_settings.py - entirely new modules, the former derived from
  Bluehound code.
- tools/config2kv.py - python script to convert file-based secrets to KV secrets.
- moved provider_settings, wsconfig, pkg_config and utility to top-level "common" folder/subpackage
- provider_settings - some minor changes to make use of the secret settings
- pkg_config - added function to return filename used for settings. moved a couple of functions to
  provider_settings.
- provider_settings - a little bit of refactoring to reduce complexity. Changed ProviderSettings class
  so that Args property is now a custom class derived from UserDict - it behaves just like a 
  dict (as it was before) but I've overridden the __getitem__ method so that it calls 
  `SecretsClient.read_secret`
- utility - added 
   - isipython() - returns true if the code is running in a notebook
   - set_unit_testing() and unit_testing() - the former sets a flag (an env var) that can be called from 
     unit testing code. The latter is meant for production code - allows you to avoid things like
     `input()` or other things that read from stdin, which cause unit tests to fail.

```
          if unit_testing():
               input("answer me"))
```
- ws_config - no real changes I think
- moving things to msticpy.common - this has caused minor changes in a lot of files including
  documentation and doc notebooks. I've kept the references to some of these things in nbtools/
  __init__.py so that `from msticpy.nbtools import *` still imports the things it used to. I've also kept
  stub versions of a couple of the moved modules in the locations where they used to be. These stub
  modules import the content of the real modules in common. This allows 
  `from msticpy.nbtools.utility import md` to still work (although it should produce a deprecation
  warning. I did this mainly so we didn't need to go an edit the Azure-Sentinel-Notebooks content.
- Added test_provider_secrets unit test. Pretty lengthy - had to mock a lot of Azure SDK functionality.
- Updated the msticpyconfig.rst documentation to reflect new functionality and described config2kv
- Other changes to RST and IPYNB files are just for the path changes for the moved modules.

File: msticpy/common/keyvault_client.py [hunk 1]
```diff
@@ -0,0 +1,1000 @@
...
+    Permissions,
+    VaultCreateOrUpdateParameters,
+    Vault,
+)
+from IPython.display import display, HTML
+import keyring
+from keyring.errors import KeyringError
+from msrest.authentication import BasicTokenAuthentication
+from msrestazure.azure_exceptions import CloudError
+import pandas.io.clipboard as pyperclip
```
Comment (petebryan) [inline msticpy/common/keyvault_client.py]: Just interested why you aren't using pyperclip natively?

Comment (ianhelle) [reply msticpy/common/keyvault_client.py]: I just thought it was one less package to have to install.

File: msticpy/common/keyvault_client.py [hunk 2]
```diff
@@ -0,0 +1,1000 @@
...
+#                 print("missing secret from keyvault, fetching manually")
+#             self._get_token()
+#         except KeyVaultErrorException as err:
+#             if self.debug:
+#                 print("bad creds in keyvault, you gotta getem")
+#                 print("here is what went wrong: %s" % str(err))
+#             self._get_token()
+
+#     def _cache_creds_keyvault(self):
+#         self.keyvault_client.set_secret(self.secret_name, json.dumps(self.config_data))
```
[the remark this answers is not in the document]
Comment (ianhelle) [reply msticpy/common/keyvault_client.py]: I think I might need it for phase 2 - it's a bit messy I know. I will either use it or clear it out.

File: conda/conda-reqs-pip.txt [hunk 3]
```diff
@@ -2,4 +2,12 @@ Kqlmagic>=0.1.106
...
 tldextract>=2.2.2
-azure-mgmt-subscription>=0.2.0
\ No newline at end of file
+azure-core>=1.2.2
+azure-identity>=1.3.0
+azure-keyvault-secrets>=4.0.0
+azure-mgmt-subscription>=0.2.0
+azure-mgmt-keyvault>=2.0.0
+azure-mgmt-resource>=2.2.0
+azure-mgmt-subscription>=0.2.0
```
[the remark this answers is not in the document]
Comment (ianhelle) [reply conda/conda-reqs-pip.txt]: done

[the remark this answers is not in the document]
Comment (ianhelle) [reply conda/conda-reqs-pip.txt]: done

File: docs/source/getting_started/msticpyconfig.rst [hunk 4]
```diff
@@ -49,6 +54,220 @@ supports storing values as simple strings or as references to named
...
+      ResourceGroup: YourResourceGroup
+      AzureRegion: RegionToCreateKV
+      VaultName: "myvault"
+      UseKeyring: True
+      Authority: global
+      AuthnType: device
+
+Under the top level ``KeyVault`` section the following entries
+are accepted. Some of these are only required if you plan to
+use msticpy to create a new KeyVault vault.
```
[the remark this answers is not in the document]
Comment (ianhelle) [reply docs/source/getting_started/msticpyconfig.rst]: done

**Pull request merged**